---
title: "Introduction to ExprCompareR"
author: "Tanaya Datar"
date: "`r format(Sys.time(), '%d %b %Y')`"
output:
  rmarkdown::html_vignette:
    toc: true
    toc_depth: 3
vignette: >
  %\VignetteIndexEntry{Introduction_TestPackage}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
# Introduction

**ExprCompareR** is an R package designed to streamline the exploration and quantification of the relationship between RNA expression and protein expression across human tissues and conditions. Researchers studying post-transcriptional regulation currently face the time-consuming task of manually retrieving RNA-seq and protein-expression data, filtering, normalizing, computing correlations, and creating visualizations for their genes of interest. `ExprCompareR` automates and integrates these steps, providing a user-friendly workflow to quickly retrieve transcriptomic and proteomic data from public repositories such as the Genotype-Tissue Expression Project (GTEx)(Lonsdale et al., 2013) and the Human Protein Atlas (HPA)(Thul et al., 2018), compute correlations, flag outlier genes, and visualize results.

ExprCompareR provides functions for calculating and plotting the Spearman correlation of RNA-seq and protein expression data from the user's genes or tissues of interest or both. Additionally, ExprCompareR includes functionalities for detecting outlier genes i.e., genes that show large discrepancies between RNA and protein expression and for comparing protein expression between cancer and healthy tissues. 

<br><br>

An overview of the functionality is here:

 **Image 1:** Depicts the functionality of functions `compute_correlation`, `correlation_genes_tissues`, `detect_outliers`, and `cancer_protein`

Include Image + add caption as above

<br><br>

For more information, try:

```r
?ExprCompareR

```
<br><br>

This document gives a quick tour of **ExprCompareR** (version 0.1.0)  functionalities. It was written in R Markdown, using the [knitr](https://cran.r-project.org/package=knitr) package for production. See `help(package="ExprCompareR")` for further details and references provided by `citation("ExprCompareR")`.

<br><br>

To download **ExprCompareR**, use the following commands:

```r
install.packages("devtools")
library(devtools)

devtools::install_github("tanaya2026/ExprCompareR", build_vignettes = TRUE)

library(ExprCompareR)

```
To list all sample functions available in the package:
``` r
ls("package:ExprCompareR")
```

To list all sample data utilized by the package:
``` r
data(package = "ExprCompareR")
```

<br><br>

**Summary table of key functions:**

| Function                      | Description                                                                                                          |
| ----------------------------- | -------------------------------------------------------------------------------------------------------------------- |
| `compute_correlation()`       | Calculates the Spearman correlation between RNA and protein expression for a user-defined list of genes **or** tissues and generates correlation plots. `compute_correlation` acts as a wrapper function and depending on the input, it calls either `correlation_genes_only()` or `correlation_tissues_only()` internally. |
| `correlation_genes_tissues()` | Calculates the Spearman correlation between RNA and protein expression for a user-defined list of genes **and** tissues and generates correlation plots.Users can choose to visualize per gene or per tissue. `correlation_genes_tissues` acts as a wrapper function and depending on the input, it calls either `per_gene_plot()` or `per_tissue_plot()` internally.|
| `detect_outliers()`           | Identifies genes that exhibit significant discrepancies between RNA and protein expression within a selected tissue, and generates outlier plots |
| `compareCancerProtein()`      | Compares protein expression levels between cancerous and normal tissue samples for a specified cancer type, and plots genes involved.          |

<br><br>

To explore all available functions:

```r
getNamespaceExports("ExprCompareR")
```

<br><br>

# Note
ExprCompareR relies on performing API calls to HPA (Thul et al., 2018) and GTEx (Lonsdale et al., 2013) via packages, HPAanalyze(Tran et al., 2019) and gtexr(Warwick et al., 2025). These calls make the vignette building slow and network dependent. Hence, as a form of a tutorial, this vignette includes static images for reference. Tests have been conducted for these functions and examples, and they do produce accurate and expected results. After installation of the package, the user can follow the vignette to confirm. These images are found in vignettes/images of the package.

<br><br>

# Tutorial

This tutorial walks you through an example workflow demonstrating how to use the core functions in `ExprCompareR`.

## Correlation

There are two main functions for correlation analysis in `ExprCompareR` namely:

1. `compute_correlation`
2. `correlation_genes_tissues`


These functions aim to calculate the Spearman correlation between RNA and protein expression of genes in various tissue contexts.To get more information about the core correlation functions and their helper functions, the help documentation will helpful to explore:

```{r, }
library(ExprCompareR)
?ExprCompareR::compute_correlation
?ExprCompareR::correlation_genes_tissues
?ExprCompareR::correlation_genes_only
?ExprCompareR::correlation_tissues_only
?ExprCompareR::per_gene_plot
?ExprCompareR::per_tissue_plot
``` 

ExprCompareR's correlation function's depend on user input i.e. their choice of genes of interest and tissues of interest. 

* If you are interested in viewing and plotting the Spearman correlation **ONLY** for genes of interest, or **ONLY** for tissues of interest, utilize the `compute_correlation` function.

* If you are interested in viewing and plotting the Spearman correlation for **BOTH** genes of interest **AND** tissues of interest, utilize the `correlation_genes_tissues`

<br>


**Input and Output of key functions in Correlation:**

| Function                      | Input                                                                                   | Output                                                                                                      |
|------------------------------|-------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------|
| `compute_correlation()`       | A list of genes **OR** a list of tissues with length ≥ 5                                | Returns a per-gene plot if gene names are provided, or a per-tissue plot if tissue names are provided.      |
| `correlation_genes_tissues()` | A list of genes (≥ 5) **AND** a list of tissues (≥ 5) **AND** a plot type (per_gene or per_tissue) | Returns a plot object showing Spearman correlations per gene or per tissue, depending on the plot choice.  |

<br>

### Compute Correlation for Genes OR Tissues

The function `compute_correlation` serves as a wrapper to compute and visualize the Spearman correlation between RNA and protein expression for a user's gene list **OR** tissue list of interest and plots the results. It requires at least five entries in the provided list to perform the computation.


Let us consider the following example as sample input the user provides, if they have a gene list of interest:

```{r per_gene_only, eval=FALSE}
gene_list <- c("MYC", "TP53", "BRCA1", "CRP", "EGFR")
output <- compute_correlation(gene_NAMES = gene_list)
output$per_gene_plot

```
```{r,echo=FALSE, fig.cap="Spearman Correlation per gene across various tissues"}
knitr::include_graphics("images/per_gene_only.png")
```

<br><br>

**Plot Key:** <br>
Spearman correlation values are represented by color intensity. High positive correlations are indicated in dark blue, low correlations around zero are white, and high negative correlations are indicated in dark red. The shade intensity reflects the magnitude of the correlation.

<br>

**Interpretation of Plot:**<br>

This function computes the Spearman correlation of the gene list of interest across standard tissues (kidney, lung, spleen, ovary, testis, breast, cerebellum, pancreas and adrenal gland) with the X axis being the individual genes and the Y axis being the Spearman coefficient.
Higher Spearman correlation coefficients indicate stronger agreement between RNA and protein expression levels across samples. Users can use this visualization to identify tissues or genes with consistent expression trends.

--------------------------------------------------------------------------------------------------------------------

Additionally, if the user has tissue list of interest, let us consider the following example as sample input the user provides:

```{r per_tissue_only, eval=FALSE}
tissue_list = c("lung", "spleen", "liver", "ovary", "testis")
output <- compute_correlation(tissue_NAMES = tissue_list)
output$per_tissue_plot

```
```{r,echo=FALSE, fig.cap="Spearman Correlation per tissue across top expressed genes"}
knitr::include_graphics("images/per_tissue_only.png")
```

**Plot Key:** <br>
Spearman correlation values are represented by color intensity. High positive correlations are indicated in dark blue, low correlations around zero are white, and high negative correlations are indicated in dark red. The shade intensity reflects the magnitude of the correlation.

<br>

**Interpretation of Plot:**<br>

This function considers the top expressed genes in each given tissue in the user's tissue list of interest, with the X axis being the individual tissues and the Y axis being the Spearman coefficient.
Higher Spearman correlation coefficients indicate stronger agreement between RNA and protein expression levels across samples. Users can use this visualization to identify tissues or genes with consistent expression trends.

--------------------------------------------------------------------------------------------------------------------

### Compute Correlation for Genes AND Tissues

This function serves as a wrapper to compute and visualize the Spearman correlation between RNA and protein expression for a user's gene list **AND** tissue list of interest and plots the results. It requires at least five entries in the both lists to perform the computation.

<br>

It produces either: <br>
1. A per-gene plot (correlation computed across tissues), OR <br>
2. A per-tissue plot (correlation computed across genes), depending on the selected plot type.<br>

<br>

Let us consider the following example as sample input the user provides, for joint analysis of specific genes across specific tissues:

```{r both_per_gene, eval=FALSE}

tissues = c("lung", "spleen", "liver", "ovary", "testis")
genes = c("MYC", "TP53", "BRCA1", "CRP", "EGFR")

result_gene <- correlation_genes_tissues(gene_NAMES = genes,
tissue_NAMES = tissues, plot_choice = "per_gene")

result_gene$per_gene_plot
```

```{r,echo=FALSE, fig.cap="Spearman Correlation per gene across various tissues"}
knitr::include_graphics("images/both_per_gene.png")
```


**Plot Key:** <br>
Spearman correlation values are represented by color intensity. High positive correlations are indicated in dark blue, low correlations around zero are white, and high negative correlations are indicated in dark red. The shade intensity reflects the magnitude of the correlation.

<br>

**Interpretation of Plot:**<br>

This function computes the Spearman correlation of the gene list of interest and tissue list of interest. The final plot is plotted with the X axis being the individual genes and the Y axis being the Spearman coefficient.
Higher Spearman correlation coefficients indicate stronger agreement between RNA and protein expression levels across samples. Users can use this visualization to identify tissues or genes with consistent expression trends.


--------------------------------------------------------------------------------------------------------------------

Or visualize these results as a plot per tissue:

```{r both_per_tissue, eval=FALSE }

tissues = c("lung", "spleen", "liver", "ovary", "testis")
genes = c("MYC", "TP53", "BRCA1", "CRP", "EGFR")
result_gene <- correlation_genes_tissues(gene_NAMES = genes,
tissue_NAMES = tissues, plot_choice = "per_tissue")
result_gene$per_tissue_plot

```
```{r,echo=FALSE, fig.cap="Spearman Correlation per tissue across genes of interest"}
knitr::include_graphics("images/both_per_tissue.png")
```

**Plot Key:** <br>
Spearman correlation values are represented by color intensity. High positive correlations are indicated in dark blue, low correlations around zero are white, and high negative correlations are indicated in dark red. The shade intensity reflects the magnitude of the correlation.

<br>

**Interpretation of Plot:**<br>
This function computes the spearman correlation of the gene list of interest and tissue list of interest. The final plot is plotted with the X axis being the individual tissues and the Y axis being the spearman coefficient.
Higher Spearman correlation coefficients indicate stronger agreement between RNA and protein expression levels across samples. Users can use this visualization to identify tissues or genes with consistent expression trends.

<br>

--------------------------------------------------------------------------------------------------------------------

## Outliers

`detect_outliers` is the core function for outlier analysis in `ExprCompareR`. <br>

This function detects outlier genes present in a given tissue of interest, which have a large delta between their RNA expression and protein expression values.The function, given a tissue of interest, compares the RNA expression and protein expression of all genes present in the given tissue, and identifies genes which have a large difference in their RNA and protein expression. These genes are identified as outlier genes, and are most likely candidates for post-transcriptional modifications. This function returns the list of outlier genes as well as a plot highlighting the outlier genes.


<br>
**Input and Output of the key function in Outliers:**

| Function                      | Input                                                                                   | Output                                                                                                      |
|------------------------------|-------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------|
| `detect_outliers()`       | The tissue of interest to analyze| Returns a plot highlighting the outlier genes in red and a vector of outlier gene names detected |


<br>

To get more information about the outlier function, explore the help documentation:

```{r, }
library(ExprCompareR)
?ExprCompareR::detect_outliers
``` 

<br>

ExprCompareR's outliers function depends on user input i.e. their choice of the tissue of interest. Let us consider the following example as sample input the user provides:

```{r outlier_detection, eval=FALSE }

output <- detect_outliers(input_tissue = "lung")
```

NOTE: Ignore warning from `gtexr`. Warnings will not affect final plot.


```{r outlier_detection_vector, eval=FALSE }
output$outlier_vector
 [1] "CD74"     "VIM"      "ENO1"     "ACTB"     "HSP90AB1" "HSPB1"    "ANXA10"  
 [8] "SPARC"    "PEX5L"    "FN1"      "AGMAT"    "NT5C1A"   "CTSD"     "SFTPA1"  
[15] "SRGN"     "LIN28A"   "FCRLA"    "ZNF593"   "RPS27A"   "MED12L"   "GC"      
[22] "SPINK7"   "SPHKAP"   "ADCY8"    "EEF1A1"   "GRHL3"    "RPL8"     "S100A11" 
[29] "S100A9"   "BSN"      "ZMAT4"    "SFTPC"    "SFTPB"    "PURG"     "SFTPA2"  
[36] "GJB4"     "HLA-DRB1" "S100A4"   "PSAP"     "S100A6"   "HLA-DRA"  "AGER"    
[43] "HLA-E"    "FAM221B"  "XKR9"     "HLA-B"    "AQP1"     "HBB"      "EEF1G"
```
<br>
**Interpretation of List:**  <br>
This function provides the list of genes in the tissue of interest which are detected as outliers.These genes are most likely candidates for post-transcriptional modifications, and this list can be used to narrow down potential targets for post-transcriptional modifications. These genes can reflect regulatory mechanism of microRNA-mediated degradation, altered translation rates, or differences in protein stability.
<br>

```
output$outlier_plot

```

```{r,echo=FALSE, fig.cap="Outlier Detection Plot"}
knitr::include_graphics("images/outlier_detection.png")
```
<br>
**Plot Key:** <br>
The black dots indicate genes which do not have a significant change in their RNA vs protein expression, whereas the red dots indicate outlier genes.<br>
These genes may reflect post-transcriptional regulation, tissue-specific translation efficiency, or data outliers.

<br>

**Interpretation of Plot:**  <br>

The scatter plot generated, depicts the outliers detected in the tissue of interest i.e. those genes who have a large delta in their RNA and protein expression. The X axis represents the RNA expression of the genes found in the tissue of interest and the Y axis represents the protein expression of genes found in the tissue of interest. These RNA values are log transformed (log 2) as they are far right skewed as compared to the protein values to have a fair comparison of RNA and protein values as they are obtained from different sources. This function utlizies IQR thresholds to identify outlier genes.

--------------------------------------------------------------------------------------------------------------------
## Pathology

`compareCancerProtein` is the core function for protein pathology analysis in `ExprCompareR`. <br>

This function compares the protein expression levels between normal and pathology (cancer) tissues for a specified cancer type. It calculates the change in expression rank and direction ('Up', 'Down', or 'No change') for each gene, and visualizes the shift in expression.
<br>

**Input and Output of the key function in Pathology:**

| Function                      | Input                                                                                   | Output                                                                                                      |
|------------------------------|-------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------|
| `compareCancerProtein`       | The cancer type to analyze | Returns a plot showing the direction and magnitude of protein rank changes for each gene and a metadata tibble with columns: <br><br> {ensembl}: Ensembl gene identifier <br> {gene}: Gene symbol <br>  {normal_rank}: Average protein expression rank in normal tissue <br> {cancer_rank}: Weighted protein expression rank in cancer tissue <br> {delta_rank}: Difference (cancer_rank - normal_rank) <br>  {direction}: Direction of change ('Up', 'Down', or 'No change')|

<br>

To get more information about the pathology function, explore the help documentation:

```{r,}
library(ExprCompareR)
?ExprCompareR::compareCancerProtein

``` 
<br>

ExprCompareR's pathology function depends on user input i.e. their choice of the cancer of interest. <br>
Let us consider the following example as sample input the user provides:

```{r pathology,eval=FALSE}

output <- compareCancerProtein(cancer_type = "breast cancer")
output$table
# A tibble: 5,378 × 6
   ensembl         gene     normal_rank cancer_rank delta_rank direction
   <chr>           <chr>          <dbl>       <dbl>      <dbl> <chr>    
 1 ENSG00000000003 TSPAN6          1          1.75       0.75  Up       
 2 ENSG00000000419 DPM1            2          2          0     No change
 3 ENSG00000000457 SCYL3           1.67       0.583     -1.08  Down     
 4 ENSG00000000460 C1orf112        2          1.67      -0.333 Down     
 5 ENSG00000001167 NFYA            1.33       2          0.667 Up       
 6 ENSG00000001461 NIPAL3          1.33       1.83       0.5   Up       
 7 ENSG00000001629 ANKIB1          1.67       1.83       0.167 Up       
 8 ENSG00000002330 BAD             1.33       2.08       0.75  Up       
 9 ENSG00000002549 LAP3            1.67       2.1        0.433 Up       
10 ENSG00000002746 HECW1           2          2.5        0.5   Up       
# ℹ 5,368 more rows
# ℹ Use `print(n = ...)` to see more rows
```
<br>
**Interpretation of Table:**<br>
As mentioned above in the output section, each table is structured, to allow researchers to investigate the role of each gene involved in this cancer type. Positive `delta_rank` values indicate up regulation in cancer, while negative values indicate down regulation.
This table assists in scientists narrowing down their gene of interest based on the delta rank and direction column. These genes maybe be affected during cancer development and can be potential driver genes.

```
output$plot

```
```{r,echo=FALSE, fig.cap="Differential Expression in Protein Plot"}
knitr::include_graphics("images/pathology.png")
```
<br>
**Interpretation of Plot:**<br>

Each gene is plotted to show the magnitude and direction of change in protein rank between normal and cancer tissue. The X axis represents the delta rank i.e. the change in protein expression of normal vs pathology. A decrease in rank is indicated by the colour red, an increase is indicated in the colour blue, and no change is indicated by green. This plot gives researchers a quick overview of the general direction of the genes in this cancer type.

<br>

# Package Reference

Datar, T. (2025) ExprCompareR:Explore and quantify the relationship between RNA-seq and protein expression across tissues and conditions.
     Unpublished. URL https://github.com/tanaya2026/ExprCompareR.

<br>

# References

* BioRender.com. BioRender [Online]. Available at: https://www.biorender.com (accessed 2 November 2025).

* Lonsdale, J., Thomas, J., Salvatore, M. et al. The Genotype-Tissue Expression (GTEx) project. Nat Genet 45, 580–585 (2013). https://doi.org/10.1038/ng.2653

* Müller K, Wickham H (2025). tibble: Simple Data Frames. R package version 3.3.0, https://tibble.tidyverse.org/.

* OpenAI.(2025). ChatGPT (GPT-5) [Large language model]. Retrieved November 2, 2025, from https://chatgpt.com/

* R Core Team (2024). _R: A Language and Environment for Statistical Computing_. R Foundation for Statistical Computing, Vienna, Austria. <https://www.R-project.org/>.

* Slowikowski K (2024). ggrepel: Automatically Position Non-Overlapping Text Labels with 'ggplot2'. https://ggrepel.slowkow.com/,https://github.com/slowkow/ggrepel.

* Thul, P. J., & Lindskog, C. (2018). The human protein atlas: A spatial map of the human proteome. Protein science : a publication of the Protein Society, 27(1), 233–244. https://doi.org/10.1002/pro.3307

* Tran AN, Dussaq AM, Kennell Jr T, Willey CD, Hjelmeland AB (2019). “HPAanalyze: an R package that facilitates the retrieval and analysis of the Human Protein Atlas data.”
MC Bioinformatics 20, 463 (2019). https://doi.org/10.1186/s12859-019-3059-z

* Warwick A, Zuckerman B, Ung C, Luben R, Olvera-Barrios A (2025). “gtexr: A convenient R interface to the Genotype-Tissue Expression (GTEx) Portal API.” Journal of Open Source Software, 10(109), 8249. ISSN 2475-9066, doi:10.21105/joss.08249, gigs v0.2.1. 

* Wickham H (2016). ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag New York. ISBN 978-3 319-24277-4, https://ggplot2.tidyverse.org. 

* Wickham H, François R, Henry L, Müller K, Vaughan D (2025). dplyr: A Grammar of Data Manipulation. R package version 1.1.4, https://dplyr.tidyverse.org.

* Wickham H, Henry L (2025). purrr: Functional Programming Tools. R package version 1.1.0, https://purrr.tidyverse.org/.

* Wickham H, Vaughan D, Girlich M (2025). tidyr: Tidy Messy Data. R package version 1.3.1, https://tidyr.tidyverse.org.


```{r}
sessionInfo()
```
